File: TopCoder/python/misc/RemoveNthNodeFromEndOfList.py

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def removeNthFromEnd(self, head, n):
        start = head
        end = head
        count = 0

        while count < n:
            end = end.next
            count = count + 1

            if end is None:
                break

        if count < n:
            return head

        if end is None:
            return start.next

        while end is not None and end.next is not None:
            start = start.next
            end = end.next

        start.next = start.next.next

        return head


if __name__ == '__main__':

    head = ListNode(1)

    sol = Solution()
    res = sol.removeNthFromEnd(head, 1)
    print(res)

File: TopCoder/src/main/java/com/leetcode/linkedlist/RemoveNthNodeFromEndOfList.java
package com.leetcode.linkedlist;

//https://leetcode.com/problems/remove-nth-node-from-end-of-list/
public class RemoveNthNodeFromEndOfList {
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }

    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null) {
            return null;
        }
        ListNode start = head;
        ListNode end = head;
        int count = 0;

        while (count < n) {
            end = end.next;
            count++;
            if (end == null) {
                break;
            }
        }

        if (count < n) {
            return start;
        }


        while (end != null && end.next != null) {
            start = start.next;
            end = end.next;
        }

        start.next = start.next.next;
        return head;

    }

    public static void main(String[] args) {
        ListNode start = new ListNode(1);
        start.next = new ListNode(2);
        start.next.next = new ListNode(3);
        start.next.next.next = new ListNode(4);
        start.next.next.next.next = new ListNode(5);
        ListNode result = new RemoveNthNodeFromEndOfList().removeNthFromEnd(start, 2);
        System.out.println(result.val);
    }
}
